### PR TITLE
Add diffusers integration for NVFP4 quantization

### PR DESCRIPTION
**Pull request by awarebayes**
# Add Hugging Face Diffusers integration for NVFP4 quantization

comparison_grid

## Summary

This PR adds first-class support for [Hugging Face Diffusers](https://github.com/huggingface/diffusers), enabling any diffusers model to be loaded with NVFP4 quantization via `from_pretrained(quantization_config=...)`. It also fixes a bug in `FourOverSixLinear` that caused crashes with models whose activations are non-contiguous.

## Motivation

fouroversix currently supports `transformers` models, but a growing number of state-of-the-art image generation models (FLUX, Stable Diffusion 3, PixArt, etc.) are distributed as diffusers pipelines. This PR bridges that gap -- users can now quantize any diffusers model to NVFP4 with a single config object, no boilerplate needed.

### New: `src/fouroversix/diffusers/` package (458 lines)

| File | Description |
|---|---|
| `config.py` | `FourOverSixConfig` -- a `QuantizationConfigMixin` subclass that mirrors `ModuleQuantizationConfig` fields and converts to fouroversix's internal config types |
| `quantizer.py` | `FourOverSixQuantizer` -- a `DiffusersQuantizer` subclass that handles module replacement, weight quantization, and pre-quantized checkpoint loading |
| `utils.py` | `replace_with_fouroversix_linear()` -- recursive `nn.Linear` → `FourOverSixLinear` replacement helper |
| `__init__.py` | Exports + auto-registration into diffusers' `AUTO_QUANTIZER_MAPPING` at import time |

### Modified existing files

- **`src/fouroversix/__init__.py`** -- Optionally exports `FourOverSixConfig` from the main package
- **`src/fouroversix/model/modules/linear.py`** -- Adds `input = input.contiguous()` before the CUDA quantize kernel call in `FourOverSixLinearFunction.forward`, fixing a crash with models that produce non-contiguous activation tensors (e.g. FLUX)
- **`pyproject.toml`** -- Adds `[project.optional-dependencies].diffusers` group (`accelerate>=0.26.0`, `diffusers>=0.32.0`) and adds `diffusers` to the `evals` group
- **`README.md`** -- Adds "Quantize a Diffusers Model to NVFP4" section with usage examples

### New scripts: `scripts/diffusers/`

| Script | Purpose |
|---|---|
| `test_diffusers.py` | Validates both `quantize_model()` and `from_pretrained()` paths on PixArt-Sigma |
| `generate_flux_klein.py` | End-to-end image generation with FLUX.2-klein-4B + NVFP4 |
| `bench_flux_klein.py` | BF16 vs NVFP4 benchmark (timing + VRAM) |

## Usage

```bash
pip install "fouroversix[diffusers]" --no-build-isolation
```

```python
import torch
from diffusers import Flux2KleinPipeline, Flux2Transformer2DModel
from fouroversix.diffusers import FourOverSixConfig

# Load transformer with NVFP4 quantization
transformer = Flux2Transformer2DModel.from_pretrained(
    "black-forest-labs/FLUX.2-klein-4B",
    subfolder="transformer",
    quantization_config=FourOverSixConfig(),
    torch_dtype=torch.bfloat16,
)

# Build pipeline with quantized transformer
pipe = Flux2KleinPipeline.from_pretrained(
    "black-forest-labs/FLUX.2-klein-4B",
    transformer=transformer,
    torch_dtype=torch.bfloat16,
)
pipe.to("cuda")

image = pipe("A cat in a spacesuit", num_inference_steps=4).images[0]
```

## Benchmark: FLUX.2-klein-4B on RTX 5090

| Metric | BF16 | NVFP4 | Improvement |
|---|---|---|---|
| Avg generation (4 steps, 1024x1024) | 0.998s | 0.630s | **1.58x faster** |
| Peak VRAM allocated | 18.60 GB | 13.04 GB | **30% less** |
| Peak VRAM reserved | 21.06 GB | 15.53 GB | **26% less** |

Image quality is visually comparable between BF16 and NVFP4:

## Design decisions

1. **Dynamic auto-quantizer registration**: diffusers uses a hardcoded `AUTO_QUANTIZER_MAPPING` dict. Rather than requiring users to manually register, the `fouroversix.diffusers` package patches the mapping at import time, so `quantization_config=FourOverSixConfig()` "just works" with any `ModelMixin.from_pretrained()` call.

2. **`_QuantMethodStr`**: diffusers expects `quant_method.value` (enum-style access) internally. Instead of forking diffusers or monkey-patching, a minimal `str` subclass with a `.value` property provides compatibility without any upstream changes.

3. **Two loading modes**: The quantizer supports both quantize-on-load (for fresh models) and pre-quantized loading (for saved NVFP4 checkpoints), matching the pattern established by diffusers' bitsandbytes integration.

4. **Contiguity fix**: The CUDA `quantize_to_fp4` kernel requires last-dim contiguity. FLUX and other architectures can produce non-contiguous activations through view/transpose ops. The one-line fix in `FourOverSixLinear.forward` ensures correctness for all models.

## Test plan

- [x] `scripts/diffusers/test_diffusers.py` -- validates `quantize_model()` and `from_pretrained()` on PixArt-Sigma, checks output shapes and dtypes
- [x] `scripts/diffusers/generate_flux_klein.py` -- end-to-end image generation with FLUX.2-klein-4B
- [x] `scripts/diffusers/bench_flux_klein.py` -- timing and VRAM comparison (BF16 vs NVFP4)
- [x] `ruff check` passes on all new and modified files

### CODE REVIEW

**Comment by awarebayes** [comment]
Updated

**Comment by jackcook** [comment]
Thank you for the contribution, and my apologies about the slow reply!
